### ISSUE TITLE
what is the road map for user space docker?

### ISSUE THREAD

**Issue by davidroberson** [rabix/bunny#385]
https://github.com/common-workflow-language/common-workflow-language/wiki/Userspace-Container-Review

Thanks

Dave

**Comment by milos-ljubinkovic** [comment]
Hi, there has been some discussion about using rabix with slurm that will most likely use something similar, but it is currently just in a planning phase.

I've made a quick proof of concept on this [branch](https://github.com/rabix/bunny/tree/task/userdocker). I've added two configs:

```
executor.override=/usr/local/bin/docker run {-v|vols} {-e|envs} {-w|workdir} {image} {command}
executor.override.pull=/usr/local/bin/docker pull {image};
```

The first one is the command used for execution, second one to pull the docker image. The first part of both can be replaced with udocker.py file or a similar implemetation. I've tested this on linux with the example dna2protein wf and it completes successfully without having root. Probably will run the conformance tests as well after some optimizing.

I've also added some basic command building in the case that user space docker implementation requires some alterations to the regular docker run command. The parameters in config between brackets will be replaced with equivalent values from regular run command (the value between the first bracket and | being used as prefix) and anything outside brackets will be verbatim used in the execution command so that udocker's execution mode and other parameters can be declared.

Current issues with this branch:

- Image pull is called every time: This can be easily removed but we will most likely want a separate and configurable image pull and installation
- there's no parallelization in running jobs: Commands are called one at a time, will easily be fixed with some checks for running processes
- there might be some issues with weird command lines: haven't tested the branch enough, previous changes to execution usually resulted in issues
